registry/handlers/blobupload.go

```diff
@@ -0,0 +1,355 @@
+			})
+		}
+		buh.Upload = upload
+
+		if state.Offset > 0 {
+			// Seek the blob upload to the correct spot if it's non-zero.
+			// These error conditions should be rare and demonstrate really
+			// problems. We basically cancel the upload and tell the client to
+			// start over.
+			if nn, err := upload.Seek(buh.State.Offset, os.SEEK_SET); err != nil {
+				defer upload.Close()
+				ctxu.GetLogger(ctx).Infof("error seeking blob upload: %v", err)
+				return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+					w.WriteHeader(http.StatusBadRequest)
+					buh.Errors.Push(v2.ErrorCodeBlobUploadInvalid, err)
+					upload.Rollback(buh)
+				})
+			} else if nn != buh.State.Offset {
+				defer upload.Close()
+				ctxu.GetLogger(ctx).Infof("seek to wrong offest: %d != %d", nn, buh.State.Offset)
+				return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+					w.WriteHeader(http.StatusBadRequest)
+					buh.Errors.Push(v2.ErrorCodeBlobUploadInvalid, err)
+					upload.Rollback(buh)
+				})
+			}
+		}
+
+		handler = closeResources(handler, buh.Upload)
+	}
+
+	return handler
+}
+
+// blobUploadHandler handles the http blob upload process.
+type blobUploadHandler struct {
+	*Context
+
+	// UUID identifies the upload instance for the current request. Using UUID
+	// to key blob writers since this implementation uses UUIDs.
+	UUID string
+
+	Upload distribution.BlobWriter
+
+	State blobUploadState
+}
+
+// StartBlobUpload begins the blob upload process and allocates a server-side
+// blob writer session.
+func (buh *blobUploadHandler) StartBlobUpload(w http.ResponseWriter, r *http.Request) {
+	blobs := buh.Repository.Blobs(buh)
+	upload, err := blobs.Writer(buh)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError) // Error conditions here?
+		buh.Errors.Push(v2.ErrorCodeUnknown, err)
+		return
+	}
+
+	buh.Upload = upload
+	defer buh.Upload.Close()
+
+	if err := buh.blobUploadResponse(w, r, true); err != nil {
+		w.WriteHeader(http.StatusInternalServerError) // Error conditions here?
+		buh.Errors.Push(v2.ErrorCodeUnknown, err)
+		return
+	}
+
+	w.Header().Set("Docker-Upload-UUID", buh.Upload.ID())
+	w.WriteHeader(http.StatusAccepted)
+}
+
+// GetUploadStatus returns the status of a given upload, identified by id.
+func (buh *blobUploadHandler) GetUploadStatus(w http.ResponseWriter, r *http.Request) {
+	if buh.Upload == nil {
+		w.WriteHeader(http.StatusNotFound)
+		buh.Errors.Push(v2.ErrorCodeBlobUploadUnknown)
+		return
+	}
+
+	// TODO(dmcgowan): Set last argument to false in blobUploadResponse when
+	// resumable upload is supported. This will enable returning a non-zero
+	// range for clients to begin uploading at an offset.
+	if err := buh.blobUploadResponse(w, r, true); err != nil {
+		w.WriteHeader(http.StatusInternalServerError) // Error conditions here?
+		buh.Errors.Push(v2.ErrorCodeUnknown, err)
+		return
+	}
+
+	w.Header().Set("Docker-Upload-UUID", buh.UUID)
```

Comment:
Please see [RFC6648](http://tools.ietf.org/html/rfc6648). ;)

I'm glad I get to tell everyone about this today!